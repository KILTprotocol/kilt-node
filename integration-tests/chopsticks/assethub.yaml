endpoint:
  - wss://rococo-asset-hub-rpc.polkadot.io
mock-signature-host: true
db: db/asset-hub.sqlite

import-storage:
  System:
    Account:
      -
        -
          - 5DPiZzQQdoJJucxGMCgrJEdeUkLfPs6fndeCMA1E4ZgAkWyh
        - providers: 1
          data:
            free: '100000000000000000000000'
  foreignAssets:
    asset:
      -  [ [{"parents":2,"interior":{"X2":[{"GlobalConsensus":{"Ethereum":{"chainId":11155111}}},{"AccountKey20":{"network":null,"key":"0xc9f05326311bc2a55426761bec20057685fb80f7"}}]}}], {"owner":"5GjRnmh5o3usSYzVmsxBWzHEpvJyHK4tKNPhjpUR3ASrruBy","issuer":"5DPiZzQQdoJJucxGMCgrJEdeUkLfPs6fndeCMA1E4ZgAkWyh","admin":"5DPiZzQQdoJJucxGMCgrJEdeUkLfPs6fndeCMA1E4ZgAkWyh","freezer": "5DPiZzQQdoJJucxGMCgrJEdeUkLfPs6fndeCMA1E4ZgAkWyh", "supply": "4242424242424242424242", "deposit": 100000000000,"minBalance": 1,"isSufficient":false,"accounts": 1,"sufficients":0,"approvals":0,"status":"Live"}]