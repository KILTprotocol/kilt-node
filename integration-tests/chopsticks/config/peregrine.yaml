endpoint:
  - wss://peregrine.kilt.io
mock-signature-host: true
db: db/peregrine.sqlite
runtime-log-level: 5
wasm-override: /home/antonio/Developer/kilt-node/target/debug/wbuild/peregrine-runtime/peregrine_runtime.compact.compressed.wasm
