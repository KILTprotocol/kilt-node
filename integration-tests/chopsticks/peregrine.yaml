db: './db/peregrine.db.sqlite'
port: 9004
endpoint: 'wss://peregrine.kilt.io'
wasm-override: ./peregrine_runtime.compact.compressed.wasm

import-storage:
    assetSwitchPool1:
        switchPair:
            # Pool account id Charlie
            poolAccountId: '5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y'
            remoteAssetBalance: '100000000000000000'
            remoteAssetSovereignTotalBalance: '90000000000000000000'
            remoteAssetCirculatingSupply: '10000000000000000000'
            remoteAssetTotalSupply: '100000000000000000000'
            remoteAssetId:
                V3:
                    Concrete:
                        parents: 2
                        interior:
                            X2:
                                - GlobalConsensus:
                                      Ethereum:
                                          chainId: 11155111
                                - AccountKey20:
                                      network: null
                                      key: '0x06012c8cf97bead5deae237070f9587f8e7a266d'
            remoteXcmFee:
                V3:
                    id:
                        concrete:
                            parents: 1
                            interior: 'Here'
                    fun:
                        Fungible: 1000000000000
            remoteReserveLocation:
                V3:
                    parents: 1
                    interior:
                        X1:
                            Parachain:
                                id: 1000
            status: 'Running'

    fungibles:
        asset:
            - - - parents: 1
                  interior: 'Here'
              # Alice account
              - owner: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
                issuer: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
                admin: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
                freezer: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
                supply: '4242424242424242424242'
                deposit: 100000000000
                minBalance: 100
                isSufficient: false
                accounts: 1
                sufficients: 0
                approvals: 0
                status: 'Live'

        account:
            - - - parents: 1
                  interior: 'Here'
                  # Alice account
                - '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
              - balance: '10000000000000000'
                status: 'Liquid'
                reason: 'Consumer'
                extra: null

    System:
        Account:
            # Alice account
            - - - '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
              - providers: 1
                data:
                    free: '100000000000000000'
            # Charlie account
            - - - '5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y'
              - providers: 1
                data:
                    free: '100000000000000000'
    polkadotXcm:
        safeXcmVersion: 3
