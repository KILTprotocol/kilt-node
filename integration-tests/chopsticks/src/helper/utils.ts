import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
