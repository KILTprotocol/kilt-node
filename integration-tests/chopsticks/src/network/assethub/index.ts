import * as Main from './main_network.js'
import * as Test from './test_network.js'

export const chains = {
	main: {
		storage: Main.storage,
		getSetupOptions: Main.getSetupOptions,
		parachainInfo: Main.parachainInfo,
	},
	test: {
		storage: Test.storage,
		getSetupOptions: Test.getSetupOptions,
		parachainInfo: Test.parachainInfo,
	},
}
