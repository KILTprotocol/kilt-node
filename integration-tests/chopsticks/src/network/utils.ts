export function getSiblingLocation(paraId: number) {
	return {
		parents: 1,
		interior: {
			X1: { Parachain: paraId },
		},
	}
}

export function getParentLocation() {
	return {
		parents: 1,
		interior: 'Here',
	}
}

export function getAccountLocationV2(addr: string) {
	return {
		V2: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						network: 'Any',
						id: addr,
					},
				},
			},
		},
	}
}

export function getAccountLocationV3(addr: string) {
	return {
		V3: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						id: addr,
					},
				},
			},
		},
	}
}

export function getNativeAssetIdLocation(amount: bigint) {
	return {
		id: { Concrete: { parents: 0, interior: 'Here' } },
		fun: { Fungible: amount },
	}
}
