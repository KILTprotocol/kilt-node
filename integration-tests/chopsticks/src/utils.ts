import { Keyring } from '@polkadot/keyring'

import { AssetSwitchSupplyParameters } from './types.js'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}

	return Number(value)
}

export function getAssetSwitchParameters(
	totalSupply: bigint = BigInt('100000000000000000000'),
	ratioCirculating: bigint = BigInt(10)
): AssetSwitchSupplyParameters {
	const circulatingSupply = (totalSupply * ratioCirculating) / BigInt(100)
	const sovereignSupply = totalSupply - circulatingSupply
	return { circulatingSupply, sovereignSupply, totalSupply }
}

export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)
export const ROC = BigInt(1e12)

export const initialBalanceKILT = BigInt(100) * KILT
export const initialBalanceDOT = BigInt(100) * DOT
export const initialBalanceHDX = BigInt(100) * HDX
export const initialBalanceROC = BigInt(100) * ROC
