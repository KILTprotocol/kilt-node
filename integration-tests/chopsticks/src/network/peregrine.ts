import { SetupOption, setupContext } from '@acala-network/chopsticks-testing'

import type { Config } from './types.js'
import { ROC, initialBalanceKILT, initialBalanceROC, toNumber } from '../utils.js'
import { AssetSwitchSupplyParameters } from '../types.js'

/// Options used to create the Peregrine context
const options = {
	endpoint: process.env.PEREGRINE_WS || 'wss://peregrine.kilt.io',
	db: './db/peregrine.db.sqlite',
	port: toNumber(process.env.PEREGRINE_PORT) || 9004,
	wasmOverride: process.env.PEREGRINE_WASM_OVERRIDE || undefined,
} as SetupOption

/// Assigns the native tokens to an accounts
export function assignNativeTokensToAccounts(addr: string[], balance: bigint = initialBalanceKILT) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance } }]),
		},
	}
}

export function setSafeXcmVersion4() {
	return {
		polkadotXcm: {
			safeXcmVersion: 4,
		},
	}
}

export function createAndAssignRocs(manager: string, addr: string[], balance: bigint = initialBalanceROC) {
	return {
		fungibles: {
			asset: [
				[
					[ROC_LOCATION],
					{
						owner: manager,
						issuer: manager,
						admin: manager,
						freezer: manager,
						supply: balance * BigInt(addr.length),
						deposit: 0,
						minBalance: 1,
						isSufficient: false,
						accounts: addr.length,
						sufficients: 0,
						approvals: 0,
						status: 'Live',
					},
				],
			],
			account: addr.map((acc) => [
				[{ parents: 1, interior: 'here' }, acc],
				{ balance: balance, status: 'Liquid', reason: 'Consumer', extra: null },
			]),
		},
	}
}

export function setSudoKey(sudoKey: string) {
	return {
		Sudo: {
			key: sudoKey,
		},
	}
}

export function setSwitchPair(
	parameters: AssetSwitchSupplyParameters,
	remoteAssetId: Record<string, unknown>,
	remoteXcmFee: Record<string, unknown>,
	remoteReserveLocation: Record<string, unknown>,
	poolAccountId: string = initialPoolAccountId,
	status: 'Running' | 'Paused' = 'Running'
) {
	return {
		assetSwitchPool1: {
			SwitchPair: {
				poolAccountId,
				remoteAssetSovereignTotalBalance: parameters.sovereignSupply,
				remoteAssetCirculatingSupply: parameters.circulatingSupply,
				remoteAssetTotalSupply: parameters.totalSupply,
				remoteAssetId,
				remoteXcmFee,
				remoteReserveLocation,
				status,
			},
		},
		// the pool account needs at least as much fund to cover the circulating supply. Give him exactly that amount + ED.
		System: {
			Account: [[[poolAccountId], { providers: 1, data: { free: parameters.circulatingSupply } }]],
		},
	}
}

/// Peregrine ParaId
export const paraId = 2086
export const PILT = { Concrete: { parents: 0, interior: 'Here' } }
export const ROC_LOCATION = {
	parents: 1,
	interior: 'Here',
}
// 0.1 ROC as remote fee
export const remoteFee = ROC / BigInt(10)

/// Sibling sovereign account for other chains
export const sovereignAccountAsSibling = '5Eg2fnshxV9kofpcNEFE7azHLAjcCtpNkbsH3kkWZasYUVKs'
// ED on Peregrine
export const existentialDeposit = BigInt('10000000000000')

export const initialPoolAccountId = '4nv4phaKc4EcwENdRERuMF79ZSSB5xvnAk3zNySSbVbXhSwS'

export async function getContext(): Promise<Config> {
	return setupContext(options)
}

export const treasuryAccount = '5EYCAe5hitmhtZcXaCrygrqw4kbmJ8bGm15PBuwUdDdQTUgY'
export const parachainStakingRewards = BigInt('760514107536694')
