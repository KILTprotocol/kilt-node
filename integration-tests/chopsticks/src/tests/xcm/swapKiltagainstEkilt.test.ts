import { test } from 'vitest'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'

import * as PeregrineConfig from '../../network/peregrine.js'
import * as AssetHubConfig from '../../network/assethub.js'
import { initialBalanceKILT, initialBalanceROC, keysAlice, keysCharlie } from '../../utils.js'
import {
	peregrineContext,
	getFreeBalancePeregrine,
	getFreeRocPeregrine,
	getFreeEkiltAssetHub,
	assethubContext,
	getFreeRocAssetHub,
	getRemoteLockedSupply,
} from '../index.js'
import { checkBalance, createBlock, setStorage, hexAddress } from '../utils.js'
import { getAccountLocationV3 } from '../../network/utils.js'

test('Swap PILTs against ePILTS on AssetHub', async ({ expect }) => {
	const { checkEvents, checkSystemEvents } = withExpect(expect)

	// Assign alice some KILT and ROC tokens
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts(
			[keysAlice.address, PeregrineConfig.poolAccountId],
			initialBalanceKILT
		),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, [keysAlice.address], initialBalanceROC),
		...PeregrineConfig.setSwapPair(),
		...PeregrineConfig.setSafeXcmVersion3(),
	})

	await setStorage(assethubContext, {
		...AssetHubConfig.assignDotTokensToAccounts(
			[keysAlice.address, PeregrineConfig.siblingSovereignAccount],
			initialBalanceROC
		),
		...AssetHubConfig.createForeignAsset(keysCharlie.address, [PeregrineConfig.siblingSovereignAccount]),
	})

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, initialBalanceROC)

	// Alice should have NO eKILT on AH
	await checkBalance(getFreeEkiltAssetHub, keysAlice.address, expect, BigInt(0))

	// initial balance of the pool account and sovereign account
	const initialBalancePoolAccount = await getFreeBalancePeregrine(PeregrineConfig.poolAccountId)
	const initialBalanceSovereignAccount = await getFreeEkiltAssetHub(PeregrineConfig.siblingSovereignAccount)
	const initialBalanceRocSovereignAccount = await getFreeRocAssetHub(PeregrineConfig.siblingSovereignAccount)
	const initialRemoteLockedSupply = await getRemoteLockedSupply()

	// 50 PILTS
	const balanceToTransfer = BigInt('50000000000000000')

	const beneficiary = getAccountLocationV3(hexAddress(keysAlice.address))

	const signedTx = peregrineContext.api.tx.assetSwap.swap(balanceToTransfer.toString(), beneficiary)

	const events = await sendTransaction(signedTx.signAsync(keysAlice))

	// Check sender state
	await createBlock(peregrineContext)

	// Check events sender
	checkEvents(events, 'xcmpQueue').toMatchSnapshot('sender events xcm queue pallet')
	checkEvents(events, 'assetSwap').toMatchSnapshot('Swap events assetSwap pallet')
	checkEvents(events, { section: 'balances', method: 'Transfer' }).toMatchSnapshot('sender events Balances')

	// check balance. Alice should have less then 50 PILTs
	const freeBalanceAlice = await getFreeBalancePeregrine(keysAlice.address)
	expect(freeBalanceAlice).toBeLessThanOrEqual(balanceToTransfer)

	// check balance Alice. Some fees should have been paid with her rocs:
	const freeRocBalanceAlice = await getFreeRocPeregrine(keysAlice.address)
	expect(freeRocBalanceAlice).eq(initialBalanceROC - BigInt(PeregrineConfig.remoteFee))

	// the swap pool account should have 50 more PILTs
	const balancePoolAccountAfterTx = await getFreeBalancePeregrine(PeregrineConfig.poolAccountId)
	expect(balancePoolAccountAfterTx).eq(initialBalancePoolAccount + balanceToTransfer)

	// Check receiver state

	await createBlock(assethubContext)

	// check events receiver
	checkSystemEvents(assethubContext, 'xcmpQueue').toMatchSnapshot('receiver events messageQueue')
	checkSystemEvents(assethubContext, { section: 'foreignAssets', method: 'Transferred' }).toMatchSnapshot(
		'receiver events Balances'
	)

	// alice should have the exact transferred amount of eKILT. Fees are paid by sovereign account
	const freeBalanceAliceAssetHub = await getFreeEkiltAssetHub(keysAlice.address)
	expect(freeBalanceAliceAssetHub).eq(balanceToTransfer)

	// sovereign account should have less eKILT by the amount of the transferred PILTs
	const freeBalanceSovereignAccount = await getFreeEkiltAssetHub(PeregrineConfig.siblingSovereignAccount)
	expect(initialBalanceSovereignAccount - balanceToTransfer).eq(freeBalanceSovereignAccount)

	// sovereign account should have paid the fees
	const freeRocsSovereignAccount = await getFreeRocAssetHub(PeregrineConfig.siblingSovereignAccount)
	expect(freeRocsSovereignAccount).toBeLessThan(initialBalanceRocSovereignAccount)

	// remote locked supply should have decreased by the amount of the transferred PILTs
	const remoteLockedSupply = await getRemoteLockedSupply()
	expect(remoteLockedSupply).eq(initialRemoteLockedSupply - balanceToTransfer)
}, 20_000)
