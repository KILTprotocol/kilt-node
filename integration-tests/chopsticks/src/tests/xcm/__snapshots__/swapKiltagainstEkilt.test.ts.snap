// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Swap PILTs against ePILTS on AssetHub > Swap events assetSwap pallet 1`] = `
[
  {
    "data": {
      "amount": 50000000000000000,
      "from": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
      "to": {
        "V3": {
          "interior": {
            "X1": {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          },
          "parents": 0,
        },
      },
    },
    "method": "LocalToRemoteSwapExecuted",
    "section": "assetSwap",
  },
]
`;

exports[`Swap PILTs against ePILTS on AssetHub > receiver events Balances 1`] = `[]`;

exports[`Swap PILTs against ePILTS on AssetHub > receiver events messageQueue 1`] = `[]`;

exports[`Swap PILTs against ePILTS on AssetHub > sender events Balances 1`] = `
[
  {
    "data": {
      "amount": 50000000000000000,
      "from": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
      "to": "4nv4phaKc4EcwENdRERuMF79ZSSB5xvnAk3zNySSbVbXhSwS",
    },
    "method": "Transfer",
    "section": "balances",
  },
]
`;

exports[`Swap PILTs against ePILTS on AssetHub > sender events xcm queue pallet 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;
